File: api/api.common/Result.fs
module Djambi.Api.Common.Result

let isOk (r : Result<'a, 'b>) =
    match r with
    | Ok _ -> true
    | _ -> false

let isError (r : Result<'a, 'b>) =
    match r with
    | Ok _ -> false
    | _ -> true

let value (r : Result<'a, 'b>) =
    match r with
    | Ok x -> x
    | _ -> raise <| invalidOp "Cannot get value of error result."

let error (r : Result<'a, 'b>) =
    match r with
    | Error x -> x
    | _ -> raise <| invalidOp "Cannot get error of ok result."

File: _archive/v3/lib/Djambi.Model/Result.cs
﻿using System;

namespace Djambi.Model
{
    public class Result<T>
    {
        private readonly T _value;

        private readonly Exception _error;

        internal Result(T value, Exception error)
        {
            _value = value;
            _error = error;
        }

        public T Value =>
            _error != null
                ? throw new InvalidOperationException($"Cannot get {nameof(Value)} from an {nameof(Error)} {nameof(Result)}.")
                : _value;

        public Exception Error =>
            _error == null
                ? throw new InvalidOperationException($"Cannot get {nameof(Error)} from a {nameof(Value)} {nameof(Result)}.")
                : _error;

        public bool HasValue => _error == null;

        public override string ToString() =>
            _error != null
                ? $"Value({_value})"
                : $"Error({_error})";
    }

    public static class Result
    { 
        public static Result<T> Value<T>(T value) =>
            new Result<T>(value, null);

        public static Result<T> Error<T>(Exception error) =>
            new Result<T>(default(T), error);
    }
}
